build/gen_stub.php

```diff
@@ -1344,70 +1417,497 @@ public function __clone()
-            if ($value === "") {
-                $code .= "\tZVAL_EMPTY_STRING(&$zvalName);\n";
-            } else {
-                $strValue = addslashes($value);
-                $code .= "\tzend_string *{$zvalName}_str = zend_string_init(\"$strValue\", sizeof(\"$strValue\") - 1, 1);\n";
-                $code .= "\tZVAL_STR(&$zvalName, {$zvalName}_str);\n";
+        $evaluator = new ConstExprEvaluator(
+            function (Expr $expr) use ($allConstInfos, &$cConstName, &$originatingConst, &$isUnknownConstValue) {
+                if ($expr instanceof Expr\ConstFetch || $expr instanceof Expr\ClassConstFetch) {
+                    if ($expr instanceof Expr\ClassConstFetch) {
+                        $constName = $expr->class->toString() . "::" . $expr->name->toString();
+                    } else {
+                        $constName = $expr->name->toString();
+                    }
+
+                    if ($constName !== "UNKNOWN") {
+                        foreach ($allConstInfos as $const) {
+                            if ($const->name->__toString() === $constName) {
+                                if ($cConstName === null) {
+                                    $cConstName = $const->cname;
+                                }
+                                if ($isUnknownConstValue === null) {
+                                    $originatingConst = $const;
+                                    $isUnknownConstValue = false;
+                                }
+
+                                return self::createFromExpression($const->value, $cConstName, $allConstInfos)->value;
+                            }
+                        }
+                    }
+
+                    if ($isUnknownConstValue === null) {
+                        $originatingConst = null;
+                        $isUnknownConstValue = true;
+                    }
+
+                    return null;
+                }
+
+                throw new Exception($this->getVariableTypeName() . " " . $this->getVariableLikeName() . " has an unsupported value");
             }
-            break;
+        );
 
-        case "array":
-            if (empty($value)) {
-                $code .= "\tZVAL_EMPTY_ARRAY(&$zvalName);\n";
-            } else {
-                throw new Exception("Unimplemented default value");
+        $result = $evaluator->evaluateDirectly($expr);
+
+        return new EvaluatedValue(
+            $result,
+            $cConstName,
+            $originatingConst,
+            (bool) $isUnknownConstValue
+        );
+    }
+
+    public static function createFromExpressionRecursively(Expr $expr, ?string $cConstName, iterable $allConstInfos): EvaluatedValue
+    {
+        $result = self::createFromExpression($expr, $cConstName, $allConstInfos);
+
+        if ($result->originatingConst !== null) {
+            $result = self::createFromExpressionRecursively($result->originatingConst->value, $result->cConstName, $allConstInfos);
```

Comment:
I'm a bit confused on why this is necessary, as `createFromExpression` already calls `createFromExpression` recursively.

Reply:
Probably the naming is misleading: `createFromExpression()` stops at the immediate value, while `createFromExpressionRecursively()` searches for the last value in the call chain if constants are referenced.

Reply:
At last, I removed the createFromExpressionRecursively method